Topics/System_Design/Design_Patterns/design_patterns.md

```diff
@@ -0,0 +1,82 @@
+
+## Model-View-Controller (MVC)
+
+### Overview
+- MVC divides an application into three interconnected components: Model, View, and Controller.
+- Model: Represents the data and business logic.
+- View: Displays the user interface.
+- Controller: Handles user input and updates the Model and View accordingly.
+
+### Use Cases
+- Web development frameworks like Ruby on Rails, Django, and ASP.NET MVC.
+- GUI applications where a clear separation between data, presentation, and user interaction is necessary.
+
+### Benefits
+- Separation of concerns enhances maintainability and scalability.
+- Promotes code organization and modularity.
+- Supports parallel development by allowing developers to work independently on different components.
+
+### Drawbacks
+- Complexity can increase with larger applications.
+- The bidirectional communication between components can lead to tight coupling.
+- Testing can be challenging due to dependencies between components.
+
+## Model-View-Presenter (MVP)
+
+### Overview
+- MVP is a derivative of MVC, focusing on improving testability and modularity.
+- Presenter acts as an intermediary between the View and the Model.
+- View is passive and does not directly interact with the Model.
+
+### Use Cases
+- Applications requiring extensive unit testing, as MVP facilitates testing by decoupling the View from the Model.
+- GUI applications where modularity and testability are critical.
+
+### Benefits
+- Enhances testability by separating concerns and reducing dependencies.
+- Facilitates code reuse and modularity.
+- Improves maintainability by isolating user interface logic from the business logic.
+
+### Drawbacks
+- Can lead to increased complexity, especially in applications with complex user interfaces.
+- Requires additional effort to implement compared to MVC.
+- Requires a clear understanding of responsibilities to avoid violating the pattern.
+
+## Model-View-ViewModel (MVVM)
+
+### Overview
+- MVVM is tailored for modern UI frameworks, such as WPF, Xamarin, and AngularJS.
+- Introduces ViewModel as an abstraction of the View's state and behavior.
+- ViewModel interacts with the Model to retrieve and manipulate data.
+
+### Use Cases
+- Data-driven applications with complex user interfaces.
+- Cross-platform development where code sharing between different platforms is essential.
+
+### Benefits
+- Enables a highly decoupled architecture, allowing for independent development of View and ViewModel.
+- Facilitates data binding, reducing boilerplate code and enhancing productivity.
+- Promotes code reusability and maintainability by encapsulating UI logic in the ViewModel.
+
+### Drawbacks
+- Learning curve for developers unfamiliar with reactive programming and data binding concepts.
+- Overuse of data binding can lead to performance issues in large-scale applications.
+- ViewModel can become bloated if not properly managed, impacting maintainability.
+
+## Conclusion
+![Patterns](DPs.png)
```

Comment:
This a great image that captures the essence of each pattern but it would be helpful if you could add explanations of what each arrow represents on the image (e.g. dotted arrow vs solid arrow vs arrow with numbers)